src/Hpack/Config.hs

```diff
@@ -271,18 +271,36 @@ instance FromJSON Condition where
 
 instance HasFieldNames Condition
 
-data ThenElse = ThenElse {
+data FlatThen a = FlatThen {
+  _flatThenCondition :: String
+, _flatThenBody :: CaptureUnknownFields (Section a)
+} deriving (Eq, Show, Generic)
+
+instance (HasFieldNames a, FromJSON a) => FromJSON (CaptureUnknownFields (FlatThen a)) where
+  parseJSON v = do
+    Condition condition <- parseJSON v
+    CaptureUnknownFields fields body <- parseJSON v
+    let unknown = fields \\ fieldNames (Proxy :: Proxy Condition)
+    return $ CaptureUnknownFields unknown (FlatThen condition (CaptureUnknownFields [] body))
+
+instance HasFieldNames a => HasFieldNames (FlatThen a) where
+  fieldNames _ = fieldNames (Proxy :: Proxy Condition) ++ fieldNames (Proxy :: Proxy a)
+
+instance (HasFieldNames a, FromJSON a) => FromJSON (FlatThen a) where
+  parseJSON v = FlatThen <$> fmap conditionCondition (parseJSON v) <*> parseJSON v
+
+data ThenElse a = ThenElse {
   _thenElseCondition :: String
-, _thenElseThen :: (CaptureUnknownFields (Section Empty))
-, _thenElseElse :: (CaptureUnknownFields (Section Empty))
+, _thenElseThen :: CaptureUnknownFields (Section a)
+, _thenElseElse :: CaptureUnknownFields (Section a)
 } deriving (Eq, Show, Generic)
 
-instance FromJSON (CaptureUnknownFields ThenElse) where
+instance (HasFieldNames a, FromJSON a) => FromJSON (CaptureUnknownFields (ThenElse a)) where
   parseJSON = captureUnknownFields
 
-instance HasFieldNames ThenElse
+instance HasFieldNames a => HasFieldNames (ThenElse a)
 
-instance FromJSON ThenElse where
+instance (HasFieldNames a, FromJSON a) => FromJSON (ThenElse a) where
   parseJSON = genericParseJSON_
 
 data Empty = Empty
```

Comment:
Cool, makes sense.

Late here already, I'll try to continue the review tomorrow (no promise).